serialization/numpy_npy.ksy

```diff
@@ -0,0 +1,232 @@
+instances:
+  data:
+    value: data_initial.data
+  dims_m_2:
+    value: parsed_header.dimensions.as<u8> - 2
+  parsed_header:
+    pos: 0
+    type: numpy_parsed_header(header)
+    doc: |
+      # save it as numpy_parsed_header.py
+      import json
+      from kaitaistruct import KaitaiStruct
+      import re
+      
+      braceFixingRx = re.compile(",\s*([\]\}])")
+      
+      def literalEvalSurrogate(s:str):
+        #A surrogate of `ast.literal_eval` as a solution for the langs not having python syntax parser
+        #BUG in KSC: `doc` cannot contain triple braces
+        s = s.strip()
+        s = s.replace("(", "[").replace(")", "]")
+        s = braceFixingRx.subn(lambda x: x.group(1), s)[0]
+        s = s.replace("True", "true").replace("False", "false")
+        s = s.replace("'", '"'))
+        return json.loads(s)
+      
+      class NumpyParsedHeader(KaitaiStruct):
+        __slots__ = ("item_type", "fortran_order", "shape", "dimensions")
+        def __init__(self, header_str: str, _io, _parent=None, _root=None):
+          self._parent = _parent
+          header = literalEvalSurrogate(header_str)
+          self.item_type = header["descr"]
+          if not isinstance(self.item_type , str):
+            raise NotImplementedError(repr(self.item_type))
+          self.fortran_order = header["fortran_order"]
+          self.shape = header["shape"]
+          self.dimensions = len(self.shape)
+types:
+  version:
+    seq:
+      - id: major
+        type: u1
+      - id: minor
+        type: u1
+  array:
+    params:
+      - id: idx
+        type: u8
+    seq:
+      - id: data
+        doc: "S<n> for n > 1 is currently out of scope"
+        type:
+          switch-on: _root.parsed_header.item_type.as<str>
+          cases:
+            '"|u1"': u1
+            '"|i1"': s1
+            '"|f1"': f1
+            '"<S1"': s1 # char
+            
+            '"<u2"': u2le
+            '"<u4"': u4le
+            '"<u8"': u8le
+            '"<i2"': s2le
+            '"<i4"': s4le
+            '"<i8"': s8le
+            '"<f2"': f2le
+            '"<f4"': f4le
+            '"<f8"': f8le
+            '"<f10"': f10le
+            '"<f12"': f12le
+            '"<f16"': f16le
+            '"<f32"': f32le
+
+            '">u2"': u2be
+            '">u4"': u4be
+            '">u8"': u8be
+            '">i2"': s2be
+            '">i4"': s4be
+            '">i8"': s8be
+            '">f2"': f2be
+            '">f4"': f4be
+            '">f8"': f8be
+            '">f10"': f10be
+            '">f12"': f12be
+            '">f16"': f16be
+            '">f32"': f32be
+
+            # machine endianness, assumming le
```

Comment:
I don't think numpy ever writes "machine-endian" dtypes into the file. When writing the dtype string, numpy looks up `.str` on the dtype object, and `numpy.dtype("|u2").str` is either `"<u2"` or `">u2"` depending on your machine's endianness (but never `"|u2"`).

Since these dtypes are ambiguous and cannot appear in npy files under normal circumstances, I think we should remove them completely instead of assuming a (possibly incorrect) byte order. That way if they do appear in a npy file for some reason, the user will get an error message instead of possibly invalid data.

Reply:
>I don't think numpy ever writes "machine-endian" dtypes into the file.

I also don't think that they appear in real files serialized by numpy. Types coded with letters  also don't appeat. But in npy spec nothing is said about that it is not allowed to use machine endianness and such npy files are really parsed correctly (on machines with the same endianness). Since the most widespread (and natural in the sense that i-th chunk uses i-th power of the base) endianness (on both desktops, mobile devices and embedded, because x86 dominates desktop, arm dominates everything else and RISCV is LE ) is currently `le` (more precisely `el8` in my notation) IMHO we should support them too.